src/pathfinding_service/service.py

```diff
@@ -143,23 +146,41 @@ def _run(self) -> None:  # pylint: disable=method-hidden
 
             self._process_new_blocks(last_block)
 
-            # Use joinall instead of sleep to collect errors from greenlets
-            gevent.joinall(
-                set([self.matrix_listener]), timeout=self._poll_interval, raise_error=True
-            )
+            # Let tests waiting for this event know that we're done with processing
+            self.updated.set()
+            self.updated.clear()
+
+            # Sleep, then collect errors from greenlets
+            gevent.sleep(self._poll_interval)
+            gevent.joinall(set([self.matrix_listener]), timeout=0, raise_error=True)
```

Comment:
What does `timeout=0`  imply?

Reply:
We don't actually wait for all greenlets (the matrix_listener won't finish unless it has an error), we just collect all greenlets that are done or threw an exception right now and then continue.

It's similar a similar concept to `gevent.sleep(0)`.